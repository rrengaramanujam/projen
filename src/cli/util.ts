import * as fs from "fs";
import * as path from "path";
import * as logging from "../logging";
import { exec } from "../util";

/**
 * Installs the npm module (through `npm install`) to node_modules under `projectDir`.
 * @param spec The npm package spec (e.g. `foo@^1.2` or `foo@/var/folders/8k/qcw0ls5pv_ph0000gn/T/projen-RYurCw/pkg.tgz`)
 * @returns The installed package name (e.g. `@foo/bar`)
 */
export function installPackage(baseDir: string, spec: string): string {
  const packageJsonPath = path.join(baseDir, "package.json");
  const packageJsonExisted = fs.existsSync(packageJsonPath);

  if (!packageJsonExisted) {
    // Make sure we have a package.json to read from later
    exec("npm init --yes", { cwd: baseDir });
  }

  logging.info(`installing external module ${spec}...`);
  exec(renderInstallCommand(baseDir, spec), { cwd: baseDir });

  // Get the true installed package name
  const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, "utf-8"));
  const packageName = Object.keys(packageJson.devDependencies).find(
    (name) => name !== "projen"
  );

  if (!packageName) {
    throw new Error(`Unable to resolve package name from spec ${spec}`);
  }

  // if package.json did not exist before calling `npm install`, we should remove it
  // so we can start off clean.
  if (!packageJsonExisted) {
    fs.rmSync(packageJsonPath, { force: true, recursive: true });
  }

  return packageName;
}

/**
 * Render a command to install an npm package.
 *
 * Engine checks are ignored at this point so that the module can be installed
 * regardless of the environment. This was needed to unblock the upgrade of the
 * minimum node version of projen, but also okay generally because engine checks
 * will be performed later and for all eternity.
 *
 * @param dir Base directory
 * @param module The module to install (e.g. foo@^1.2)
 * @returns The string that includes the install command ("npm install ...")
 */
export function renderInstallCommand(dir: string, module: string): string {
  return `npm install --save --save-dev -f --no-package-lock --prefix="${dir}" ${module}`;
}

export class CliError extends Error {
  constructor(...lines: string[]) {
    super(lines.join("\n"));
    this.name = "CliError";
  }
}
