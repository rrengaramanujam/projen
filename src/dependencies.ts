import * as path from "path";
import { PROJEN_DIR } from "./common";
import { Component } from "./component";
import { JsonFile } from "./json";
import { Project } from "./project";

/**
 * The `Dependencies` component is responsible to track the list of dependencies
 * a project has, and then used by project types as the model for rendering
 * project-specific dependency manifests such as the dependencies section
 * `package.json` files.
 *
 * To add a dependency you can use a project-type specific API such as
 * `nodeProject.addDeps()` or use the generic API of `project.deps`:
 */
export class Dependencies extends Component {
  /**
   * The project-relative path of the deps manifest file.
   */
  public static readonly MANIFEST_FILE = path.posix.join(
    PROJEN_DIR,
    "deps.json"
  );

  /**
   * Returns the coordinates of a dependency spec.
   *
   * Given `foo@^3.4.0` returns `{ name: "foo", version: "^3.4.0" }`.
   * Given `bar@npm:@bar/legacy` returns `{ name: "bar", version: "npm:@bar/legacy" }`.
   */
  public static parseDependency(spec: string): DependencyCoordinates {
    const scope = spec.startsWith("@");
    if (scope) {
      spec = spec.substr(1);
    }

    const [module, ...version] = spec.split("@");
    const name = scope ? `@${module}` : module;
    if (version.length == 0) {
      return { name };
    } else {
      return { name, version: version?.join("@") };
    }
  }

  private readonly _deps = new Array<Dependency>();

  /**
   * Adds a dependencies component to the project.
   * @param project The parent project
   */
  constructor(project: Project) {
    super(project);

    // this is not really required at the moment, but actually quite useful as a
    // checked-in source of truth for dependencies and will potentially be
    // valuable in the future for CLI tools.
    if (!project.ejected) {
      new JsonFile(project, Dependencies.MANIFEST_FILE, {
        omitEmpty: true,
        obj: () => this.toJson(),
      });
    }
  }

  /**
   * A copy of all dependencies recorded for this project.
   *
   * The list is sorted by type->name->version
   */
  public get all(): Dependency[] {
    return [...this._deps].sort(compareDeps).map(normalizeDep);
  }

  /**
   * Returns a dependency by name.
   *
   * Fails if there is no dependency defined by that name or if `type` is not
   * provided and there is more then one dependency type for this dependency.
   *
   * @param name The name of the dependency
   * @param type The dependency type. If this dependency is defined only for a
   * single type, this argument can be omitted.
   *
   * @returns a copy (cannot be modified)
   */
  public getDependency(name: string, type?: DependencyType): Dependency {
    const dep = this.tryGetDependency(name, type);
    if (!dep) {
      const msg = type
        ? `there is no ${type} dependency defined on "${name}"`
        : `there is no dependency defined on "${name}"`;

      throw new Error(msg);
    }

    return dep;
  }

  /**
   * Returns a dependency by name.
   *
   * Returns `undefined` if there is no dependency defined by that name or if
   * `type` is not provided and there is more then one dependency type for this
   * dependency.
   *
   * @param name The name of the dependency
   * @param type The dependency type. If this dependency is defined only for a
   * single type, this argument can be omitted.
   *
   * @returns a copy (cannot be modified) or undefined if there is no match
   */
  public tryGetDependency(
    name: string,
    type?: DependencyType
  ): Dependency | undefined {
    const idx = this.tryGetDependencyIndex(name, type);
    if (idx === -1) {
      return undefined;
    }

    return {
      ...normalizeDep(this._deps[idx]),
    };
  }

  /**
   * Adds a dependency to this project.
   * @param spec The dependency spec in the format `MODULE[@VERSION]` where
   * `MODULE` is the package-manager-specific module name and `VERSION` is an
   * optional semantic version requirement (e.g. `^3.4.0`).
   * @param type The type of the dependency.
   */
  public addDependency(
    spec: string,
    type: DependencyType,
    metadata: { [key: string]: any } = {}
  ): Dependency {
    this.project.logger.debug(`${type}-dep ${spec}`);

    const dep: Dependency = {
      ...Dependencies.parseDependency(spec),
      type,
      metadata,
    };

    const existingDepIndex = this.tryGetDependencyIndex(dep.name, type);

    if (existingDepIndex !== -1) {
      this.project.logger.debug(
        `updating existing ${dep.type}-dep ${dep.name} with more specific version/metadata`
      );
      this._deps[existingDepIndex] = dep;
    } else {
      this._deps.push(dep);
    }

    return dep;
  }

  /**
   * Removes a dependency.
   * @param name The name of the module to remove (without the version)
   * @param type The dependency type. This is only required if there the
   * dependency is defined for multiple types.
   */
  public removeDependency(name: string, type?: DependencyType) {
    const removeIndex = this.tryGetDependencyIndex(name, type);
    if (removeIndex === -1) {
      return;
    }

    this._deps.splice(removeIndex, 1);
  }

  private tryGetDependencyIndex(name: string, type?: DependencyType): number {
    const deps = this._deps.filter((d) => d.name === name);
    if (deps.length === 0) {
      return -1; // not found
    }

    if (!type) {
      if (deps.length > 1) {
        throw new Error(
          `"${name}" is defined for multiple dependency types: ${deps
            .map((d) => d.type)
            .join(",")}. Please specify dependency type`
        );
      }

      type = deps[0].type;
    }

    return this._deps.findIndex(
      (dep) => dep.name === name && dep.type === type
    );
  }

  private toJson(): DepsManifest | undefined {
    if (this._deps.length === 0) {
      return undefined;
    }
    return {
      dependencies: this._deps.sort(compareDeps).map(normalizeDep),
    };
  }
}

function normalizeDep(d: Dependency) {
  const obj: any = {};
  for (const [k, v] of Object.entries(d)) {
    if (v == undefined) {
      continue;
    }
    if (typeof v === "object" && Object.keys(v).length === 0) {
      continue;
    }
    if (Array.isArray(v) && v.length === 0) {
      continue;
    }
    obj[k] = v;
  }

  return obj;
}

function compareDeps(d1: Dependency, d2: Dependency) {
  return specOf(d1).localeCompare(specOf(d2));

  function specOf(dep: Dependency) {
    let spec = dep.type + ":" + dep.name;
    if (dep.version) {
      spec += "@" + dep.version;
    }
    return spec;
  }
}

export interface DepsManifest {
  /**
   * All dependencies of this module.
   */
  readonly dependencies: Dependency[];
}

/**
 * Coordinates of the dependency (name and version).
 */
export interface DependencyCoordinates {
  /**
   * The package manager name of the dependency (e.g. `leftpad` for npm).
   *
   * NOTE: For package managers that use complex coordinates (like Maven), we
   * will codify it into a string somehow.
   */
  readonly name: string;

  /**
   * Semantic version version requirement.
   *
   * @default - requirement is managed by the package manager (e.g. npm/yarn).
   */
  readonly version?: string;
}

/**
 * Represents a project dependency.
 */
export interface Dependency extends DependencyCoordinates {
  /**
   * Which type of dependency this is (runtime, build-time, etc).
   */
  readonly type: DependencyType;

  /**
   * Additional JSON metadata associated with the dependency (package manager
   * specific).
   * @default {}
   */
  readonly metadata?: { [key: string]: any };
}

/**
 * Type of dependency.
 */
export enum DependencyType {
  /**
   * The dependency is required for the program/library during runtime.
   */
  RUNTIME = "runtime",

  /**
   * The dependency is required at runtime but expected to be installed by the
   * consumer.
   */
  PEER = "peer",

  /**
   * The dependency is bundled and shipped with the module, so consumers are not
   * required to install it.
   */
  BUNDLED = "bundled",

  /**
   * The dependency is required to run the `build` task.
   */
  BUILD = "build",

  /**
   * The dependency is required to run the `test` task.
   */
  TEST = "test",

  /**
   * The dependency is required for development (e.g. IDE plugins).
   */
  DEVENV = "devenv",

  /**
   * Transient dependency that needs to be overwritten.
   *
   * Available for Node packages
   */
  OVERRIDE = "override",

  /**
   * An optional dependency that may be used at runtime if available, but is not required.
   * It is expected to be installed by the consumer.
   */
  OPTIONAL = "optional",
}
