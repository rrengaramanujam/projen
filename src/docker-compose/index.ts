export * from "./docker-compose-port";
export * from "./docker-compose-service";
export * from "./docker-compose-volume";
export * from "./docker-compose-network";
export * from "./docker-compose";
