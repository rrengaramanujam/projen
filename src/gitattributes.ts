import { IResolver, FileBase } from "./file";
import { Project } from "./project";

/**
 * Assign attributes to file names in a git repository.
 *
 * @see https://git-scm.com/docs/gitattributes
 */
export class GitAttributesFile extends FileBase {
  private readonly attributes = new Map<string, Set<string>>();

  public constructor(project: Project) {
    super(project, ".gitattributes", {
      editGitignore: false,
    });
  }

  /**
   * Maps a set of attributes to a set of files.
   * @param glob Glob pattern to match files in the repo
   * @param attributes Attributes to assign to these files.
   */
  public addAttributes(glob: string, ...attributes: string[]) {
    if (!this.attributes.has(glob)) {
      this.attributes.set(glob, new Set());
    }
    const set = this.attributes.get(glob)!;
    for (const attribute of attributes) {
      set.add(attribute);
    }
  }

  /**
   * Add attributes necessary to mark these files as stored in LFS
   */
  public addLfsPattern(glob: string) {
    this.addAttributes(glob, "filter=lfs", "diff=lfs", "merge=lfs", "-text");
  }

  /**
   * Whether the current gitattributes file has any LFS patterns
   */
  public get hasLfsPatterns() {
    return Array.from(this.attributes.values()).some((attrs) =>
      attrs.has("filter=lfs")
    );
  }

  protected synthesizeContent(_: IResolver): string | undefined {
    // We can assume the file map is never empty.
    const entries = Array.from(this.attributes.entries()).sort(([l], [r]) =>
      l.localeCompare(r)
    );

    if (entries.length === 0) {
      return undefined;
    }

    return [
      ...(this.marker ? [`# ${this.marker}`] : []),
      "",
      ...entries.map(
        ([name, attributes]) => `${name} ${Array.from(attributes).join(" ")}`
      ),
    ].join("\n");
  }
}
