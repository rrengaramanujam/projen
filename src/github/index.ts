export * from "./github";

export * from "./dependabot";
export * from "./workflows";
export * from "./actions-provider";
export * from "./mergify";
export * from "./pr-template";
export * from "./mergify";
export * from "./auto-merge";
export * from "./auto-approve";
export * from "./stale";
export * from "./pull-request-lint";
export * from "./github-credentials";
export * from "./workflow-actions";
export * from "./workflow-jobs";

export * from "./task-workflow";

export * from "./github-project";

export * as workflows from "./workflows-model";
