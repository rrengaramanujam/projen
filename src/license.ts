import * as fs from "fs";
import { FileBase, IResolver } from "./file";
import { Project } from "./project";

export interface LicenseOptions {
  /**
   * License type (SPDX).
   *
   * @see https://github.com/projen/projen/tree/main/license-text for list of supported licenses
   */
  readonly spdx: string;

  /**
   * Copyright owner.
   *
   * If the license text has $copyright_owner, this option must be specified.
   *
   * @default -
   */
  readonly copyrightOwner?: string;

  /**
   * Period of license (e.g. "1998-2023")
   *
   * The string `$copyright_period` will be substituted with this string.
   *
   * @default - current year (e.g. "2020")
   */
  readonly copyrightPeriod?: string;
}

export class License extends FileBase {
  private readonly text: string;

  constructor(project: Project, options: LicenseOptions) {
    super(project, "LICENSE", {
      marker: false,
      // Allow the license to be displayed in the repositories.
      committed: true,
    });

    const spdx = options.spdx;

    const textFile = `${__dirname}/../license-text/${spdx}.txt`;
    if (!fs.existsSync(textFile)) {
      throw new Error(`unsupported license ${spdx}`);
    }

    const years =
      options.copyrightPeriod ?? new Date().getFullYear().toString();
    const owner = options.copyrightOwner;

    let text = fs.readFileSync(textFile, "utf-8");

    text = text.replace("$copyright_period", years);

    // if the license text includes $copyright_owner, then `copyrightOwner` is required.
    if (text.indexOf("$copyright_owner") !== -1) {
      if (!owner) {
        throw new Error(
          `The ${spdx} license requires "copyrightOwner" to be specified`
        );
      }

      text = text.replace("$copyright_owner", owner);
    }

    this.text = text;
  }

  protected synthesizeContent(_: IResolver): string | undefined {
    return this.text;
  }
}
