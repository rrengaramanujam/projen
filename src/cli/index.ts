import { resolve } from "path";
import * as yargs from "yargs";
import newCommand from "./cmds/new";
import { synth } from "./synth";
import { discoverTaskCommands } from "./tasks";
import { PROJEN_DIR, PROJEN_RC, PROJEN_VERSION } from "../common";
import * as logging from "../logging";
import { TaskRuntime } from "../task-runtime";
import { findUp, getNodeMajorVersion } from "../util";

const DEFAULT_RC = resolve(PROJEN_RC);

async function main() {
  const ya = yargs;
  ya.command(newCommand);

  const pathToProjenDir = findUp(PROJEN_DIR, process.cwd());
  const runtime = new TaskRuntime(pathToProjenDir ?? ".");
  discoverTaskCommands(runtime, ya);

  ya.recommendCommands();
  ya.strictCommands();
  ya.showHelpOnFail(false);
  ya.wrap(yargs.terminalWidth());
  ya.option("post", {
    type: "boolean",
    default: true,
    desc: "Run post-synthesis steps such as installing dependencies. Use --no-post to skip",
  });
  ya.option("watch", {
    type: "boolean",
    default: false,
    desc: "Keep running and resynthesize when projenrc changes",
    alias: "w",
  });
  ya.options("debug", { type: "boolean", default: false, desc: "Debug logs" });
  ya.options("rc", {
    deprecated: true,
    desc: "path to .projenrc.js file",
    default: DEFAULT_RC,
    type: "string",
  });
  ya.completion();
  ya.help();

  // do not use the default yargs '--version' implementation since it is
  // global by default (it appears on all subcommands)
  ya.version(false);
  ya.option("version", {
    type: "boolean",
    description: "Show version number",
    global: false,
  });

  const args = await ya.argv;

  if (args.debug) {
    process.env.DEBUG = "true";
  }

  const nodeVersion = getNodeMajorVersion();
  if (nodeVersion && nodeVersion < 16) {
    logging.warn(
      `WARNING: You are using Node v${nodeVersion}, which reaches end of life on April 30, 2023. Support for EOL Node releases may be dropped by projen in the future. Please consider upgrading to Node >= 16 as soon as possible.`
    );
  }

  // no command means synthesize
  if (args._.length === 0) {
    if (args.version) {
      console.log(PROJEN_VERSION);
      process.exit(0);
    }
    await synth(runtime, {
      post: args.post as boolean,
      watch: args.watch as boolean,
      rcfile: args.rc as string,
    });
  }
}

main().catch((e) => {
  console.error(e.stack);
  process.exit(1);
});
