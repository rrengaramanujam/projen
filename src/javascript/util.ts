import { existsSync, readFileSync } from "fs";
import { basename, dirname, extname, join, sep, resolve } from "path";
import * as semver from "semver";
import { findUp } from "../util";

/**
 * Basic interface for `package.json`.
 */
interface PackageManifest {
  [key: string]: any;

  /**
   * Package name.
   */
  name: string;
  /**
   * Package version.
   */
  version: string;
}

export function renderBundleName(entrypoint: string) {
  const parts = join(entrypoint).split(sep);
  if (parts[0] === "src") {
    parts.shift(); // just remove 'src' if its the first element for ergonomics
  }

  const p = parts.join(sep);
  const dir = dirname(p);
  const base = basename(p, extname(p));
  return join(dir, base);
}

/**
 * Regex for AWS CodeArtifact registry
 */
export const codeArtifactRegex =
  /^https:\/\/(?<registry>(?<domain>[^\.]+)-(?<accountId>\d{12})\.d\.codeartifact\.(?<region>[^\.]+).*\.amazonaws\.com\/.*\/(?<repository>[^\/.]+)\/)/;

/**
 * gets AWS details from the Code Artifact registry URL
 * throws exception if not matching expected pattern
 * @param registryUrl Code Artifact registry URL
 * @returns object containing the (domain, accountId, region, repository)
 */
export function extractCodeArtifactDetails(registryUrl: string) {
  const match = registryUrl.match(codeArtifactRegex);
  if (match?.groups) {
    const { domain, accountId, region, repository, registry } = match.groups;
    return { domain, accountId, region, repository, registry };
  }
  throw new Error("Could not get CodeArtifact details from npm Registry");
}

export function minVersion(version: string): string | undefined {
  if (semver.validRange(version)) {
    return semver.minVersion(version)?.version;
  } else {
    return version;
  }
}

/**
 * Attempt to resolve location of the given `moduleId`.
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveModule(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  try {
    return require.resolve(moduleId, options);
  } catch {
    return undefined;
  }
}

/**
 * Attempt to resolve a module's manifest (package.json) path via `require.resolve` lookup.
 *
 * @remarks
 * If the target package has `exports` that differ from the default
 * (i.e, it defines the `exports` field in its manifest) and does not
 * explicitly include an entry for `package.json`, this strategy will fail.
 * See {@link tryResolveManifestPathFromDefaultExport} as an alternative.
 *
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveModuleManifestPath(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  // cannot just `require('dependency/package.json')` here because
  // `options.paths` may not overlap with this node proc's resolution paths.
  const manifestId = `${moduleId}/package.json`;
  return tryResolveModule(manifestId, options);
}

/**
 * Attempt to resolve a module's manifest (package.json) path by looking for the nearest
 * `package.json` file that is an ancestor to the module's default export location.
 *
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveManifestPathFromDefaultExport(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  const defaultExportPath = tryResolveModule(moduleId, options);
  if (!defaultExportPath) {
    return undefined;
  }
  const moduleDir = findUp("package.json", defaultExportPath);
  if (!moduleDir) {
    return undefined;
  }
  return join(moduleDir, "package.json");
}

/**
 * Attempt to resolve a module's manifest (package.json) path by checking for its existence under `node_modules` relative to `basePath`.
 *
 * @remarks
 * This strategy can be helpful in the scenario that a module defines
 * custom exports without `package.json` and no default export (i.e, some type definition packages).
 *
 * @param moduleId Module ID to lookup.
 * @param basePath Root path to search from.
 */
export function tryResolveManifestPathFromPath(
  moduleId: string,
  basePath: string
) {
  const base = basePath.includes("node_modules")
    ? basePath
    : join(basePath, "node_modules");
  const filePath = resolve(base, ...moduleId.split("/"), "package.json");
  if (existsSync(filePath)) {
    return filePath;
  }
  return undefined;
}

/**
 * Attempt to resolve a module's manifest (package.json) path by searching for it in the optionally provided paths array
 * as well as the current node processes' default resolution paths.
 * @param moduleId Module ID to search for.
 * @param options Search options.
 */
export function tryResolveManifestPathFromSearch(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  const searchPaths = [
    ...(options?.paths ?? []),
    ...(require.resolve.paths(moduleId) ?? []),
  ];
  for (const path of searchPaths) {
    const result = tryResolveManifestPathFromPath(moduleId, path);
    // early return on first result.
    if (result) {
      return result;
    }
  }
  return undefined;
}

/**
 * Attempt to resolve a module's manifest (package.json) using multiple strategies.
 * @param moduleId Module to resolve manifest path for.
 * @param options Resolution options.
 */
export function tryResolveModuleManifest(
  moduleId: string,
  options?: { paths: string[] }
): PackageManifest | undefined {
  const strategies = [
    tryResolveModuleManifestPath,
    tryResolveManifestPathFromDefaultExport,
    tryResolveManifestPathFromSearch,
  ];
  for (const strategy of strategies) {
    const result = strategy(moduleId, options);
    // early return on first result.
    if (result) {
      try {
        const manifest = JSON.parse(
          readFileSync(result, "utf8")
        ) as PackageManifest;
        // verify name matches target module.
        if (manifest.name === moduleId) {
          return manifest;
        }
      } catch {
        // continue to next strategy.
      }
    }
  }
  return undefined;
}

/**
 * Attempt to resolve the installed version of a given dependency.
 * @param dependencyName Name of dependency.
 * @param options Optional options passed through to `require.resolve`.
 */
export function tryResolveDependencyVersion(
  dependencyName: string,
  options?: { paths: string[] }
): string | undefined {
  const manifest = tryResolveModuleManifest(dependencyName, options);
  if (!manifest) {
    return undefined;
  }
  return manifest?.version;
}
