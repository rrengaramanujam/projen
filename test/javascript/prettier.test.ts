import { SourceCode } from "../../src";
import {
  ArrowParens,
  NodeProject,
  PrettierOverride,
  TrailingComma,
} from "../../src/javascript";
import { synthSnapshot } from "../util";

describe("prettier", () => {
  test("snapshot", () => {
    // GIVEN
    const project = new NodeProject({
      name: "test",
      defaultReleaseBranch: "master",
      prettier: true,
      prettierOptions: { settings: { printWidth: 140 } },
    });

    // THEN
    expect(synthSnapshot(project)[".prettierrc.json"]).toMatchSnapshot();
  });

  test("snapshot with ignore", () => {
    // GIVEN
    const project = new NodeProject({
      name: "test",
      defaultReleaseBranch: "master",
      prettier: true,
      prettierOptions: { settings: { printWidth: 140 } },
    });

    project.prettier?.ignoreFile?.addPatterns("build");

    // THEN
    expect(synthSnapshot(project)[".prettierignore"]).toMatchSnapshot();
  });

  test("sample config is created", () => {
    // GIVEN
    const project = new NodeProject({
      name: "test",
      defaultReleaseBranch: "master",
      prettier: true,
      prettierOptions: {
        settings: {
          trailingComma: TrailingComma.ALL,
          bracketSpacing: true,
          tabWidth: 2,
          semi: true,
          singleQuote: true,
          arrowParens: ArrowParens.ALWAYS,
          printWidth: 140,
          useTabs: false,
          parser: "typescript",
        },
      },
    });

    // THEN
    expect(project.prettier?.settings).toMatchObject({
      trailingComma: "all",
      bracketSpacing: true,
      tabWidth: 2,
      semi: true,
      singleQuote: true,
      arrowParens: "always",
      printWidth: 140,
      useTabs: false,
      parser: "typescript",
    });
  });

  test("snapshot with ignore", () => {
    // GIVEN
    const project = new NodeProject({
      name: "test",
      defaultReleaseBranch: "master",
      prettier: true,
    });
    new SourceCode(project, "src/example.ts");

    // THEN
    expect(synthSnapshot(project)[".prettierignore"]).toMatchSnapshot();
  });

  test("overrides", () => {
    // GIVEN
    const override: PrettierOverride = {
      files: "*.js",
      options: { parser: "typescript" },
    };
    const project = new NodeProject({
      name: "test",
      defaultReleaseBranch: "master",
      prettier: true,
    });
    project.prettier?.addOverride(override);

    // THEN
    expect(synthSnapshot(project)[".prettierrc.json"]).toHaveProperty(
      "overrides",
      [override]
    );
  });

  test("can output yml instead of json", () => {
    // GIVEN
    const project = new NodeProject({
      name: "test",
      defaultReleaseBranch: "main",
      prettier: true,
      prettierOptions: {
        yaml: true,
      },
    });

    // THEN
    const output = synthSnapshot(project);
    expect(output[".prettierrc.yml"]).toBeDefined();
    expect(output[".prettierrc.json"]).toBeUndefined();
  });
});
