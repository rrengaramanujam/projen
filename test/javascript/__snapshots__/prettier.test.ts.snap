// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prettier snapshot 1`] = `
{
  "overrides": [],
  "printWidth": 140,
}
`;

exports[`prettier snapshot with ignore 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
build
"
`;

exports[`prettier snapshot with ignore 2`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
src/example.ts
"
`;
